lib/TPP/GPU/GpuPipeline.cpp

```diff
@@ -0,0 +1,127 @@
+//===- GpuPipeline.cpp -------------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Passes.h"
+
+#include "mlir/Conversion/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/Transforms/Passes.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/LLVMIR/NVVMDialect.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Passes.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/NVGPU/IR/NVGPUDialect.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/InitAllDialects.h"
+#include "mlir/InitAllPasses.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/Passes.h"
+
+#include "TPP/Dialect/Tpp/TppDialect.h"
+#include "TPP/Dialect/Transform/LinalgXTransformOps.h"
+#include "TPP/PassUtils.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+
+using namespace mlir;
+using namespace mlir::tpp;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+
+namespace {
+
+enum class GpuType {
+  None, // no target GPU
+  Cuda,
+};
+
+GpuType parseGpuOption(StringRef gpuStr) {
+  return llvm::StringSwitch<GpuType>(gpuStr)
+      .CaseLower("none", GpuType::None)
+      .CaseLower("cuda", GpuType::Cuda)
+      .Default(GpuType::None);
+}
+
+// GPU pipeline - map and lower operations to enable execution on a GPU.
+struct GpuPipeline : public GpuPipelineBase<GpuPipeline>,
+                     UtilityPassBase<ModuleOp> {
+  GpuPipeline() = default;
+  GpuPipeline(StringRef gpuBackend) { this->gpuBackend = gpuBackend.str(); }
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<tpp::TppDialect>();
+    registry.insert<linalg::LinalgDialect>();
+    registry.insert<scf::SCFDialect>();
+    registry.insert<memref::MemRefDialect>();
+    registry.insert<gpu::GPUDialect>();
+    registry.insert<LLVM::LLVMDialect>();
+    registry.insert<NVVM::NVVMDialect>();
+    registry.insert<nvgpu::NVGPUDialect>();
+  }
+
+  void runOnOperation() override {
+    auto module = getOperation();
+
+    // Initialize the pipeline if needed.
+    // Otherwise, just run the cached one.
+    if (pm.empty())
+      constructPipeline();
+
+    if (failed(runPipeline(pm, module)))
+      return signalPassFailure();
+  }
+
+private:
+  void constructPipeline() override {
+    pm.clear();
+
+    GpuType gpuType = parseGpuOption(this->gpuBackend);
+
+    // Add no passes when no GPU is selected.
+    if (gpuType == GpuType::None)
+      return;
```

Comment:
What is the actual case when someone selects `USE_GPU=1` and `gpu_type=none`?

Do we even need that option?

Reply:
You can enable GPU support but choose to still target only CPU.
The explicit `none` is a bit silly so it is removed now.